iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -388,53 +484,68 @@ public override async Task OpenAsync(CancellationToken cancellationToken)
             // The current behavior does not support open->close->open
             if (_isDisposed)
             {
-                throw new ObjectDisposedException(nameof(RetryDelegatingHandler));
+                throw new ObjectDisposedException("IoT client", ClientDisposedMessage);
             }
-
-            if (_isOpen)
+            switch (GetClientTransportStatus())
             {
-                return;
-            }
-
-            await _handlerSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
-            try
-            {
-                if (!_isOpen)
-                {
-                    if (Logging.IsEnabled)
-                        Logging.Info(this, "Opening connection", nameof(OpenAsync));
+                case ClientTransportStatus.Open:
+                case ClientTransportStatus.Opening:
+                    return;
 
-                    // This is to ensure that if OpenInternalAsync() fails on retry expiration with a custom retry policy,
-                    // we are returning the corresponding connection status change event => disconnected: retry_expired.
-                    try
-                    {
-                        await OpenInternalAsync(cancellationToken).ConfigureAwait(false);
-                    }
-                    catch (Exception ex) when (!Fx.IsFatal(ex))
-                    {
-                        HandleConnectionStatusExceptions(ex, true);
-                        throw;
-                    }
+                case ClientTransportStatus.Closing:
+                    throw new InvalidOperationException($"The client is currently closing. Wait until {nameof(CloseAsync)} completes and then invoke {nameof(OpenAsync)} again.");
```

Comment:
If `Openasync()` is called while the client is closing, throw an InvalidOperationException.

Reply:
Updating the formatting marked this comment as outdated. Resolving this and adding a new one.